### ISSUE TITLE
Improve Heap Allocation In Index::blocks

### ISSUE THREAD

**Issue by bonedaddy** [ordinals/ord#1893]
# Overview

Heap allocation is inefficient in `index::blocks`, as it initializes a vector with an empty capacity and pushes elements into the vec. This will result in numerous allocations being performed, when the function can be optimized to perform only one allocation:

# Solution

Initialize the vector with a fixed capacity after retrieving `block_count`. If `block_count` is `100` this would avoid 6 total allocations in favour of one allocation (note: rust reallocs vector size by 2x). With the current bitcoin blockchain size, this would reduce hundreds of allocations into one allocation leading to drastically improved performance.

```rust
    let rtx = self.begin_read()?;

    let block_count = rtx.block_count()?;

    let height_to_block_hash = rtx.0.open_table(HEIGHT_TO_BLOCK_HASH)?;

    let mut blocks = Vec::with_capacity(block_count);

    for next in height_to_block_hash.range(0..block_count)?.rev().take(take) {
      blocks.push((next.0.value(), Entry::load(*next.1.value())));
    }
```

**Comment by veryordinally** [comment]
Makes sense. Have you tried this and benchmarked?

**Comment by bonedaddy** [comment]
This is from using both improvements introduced in https://github.com/casey/ord/pull/1959 and https://github.com/casey/ord/pull/1960 but pretty fast speed boost
![2023-03-21_18-19](https://user-images.githubusercontent.com/17089485/226777383-87e7fac8-1f44-467a-8664-4a9cae15621c.png)

**Comment by veryordinally** [comment]
Sounds great - just to make sure I understand your benchmark data: this is timing for an index run with your optimizations? What was the time without the optimizations? Would be great to get a rough idea of speedup you see. Absolute times are hard to compare across setups.

**Comment by cryptoni9n** [comment]
This appears to have been merged in #2409 and should therefore be closed.
